### PR TITLE
[VTAdmin] Dynamic Clusters

### PR DESCRIPTION
**Pull request by notfelineit**
Allow clusters in VTAdmin to be selected (and filtered down) dynamically - vtadmin-web will specify which clusters it "wants" and vtadmin-api will only work on those clusters.

## Description
This PR modifies vtadmin-api slightly to support dynamic clusters - clusters that are specified by the client (vtadmin-web) via cookie parameters. If `http-enable-dynamic-clusters` is not set, behavior **remains the same**. If it is set to true, the following changes apply:
- `http-enable-dynamic-clusters` flag is introduced
- Dynamic routes: the `/api` routes are dynamic in that the clusters in the `api` object are filtered down to only the one specified by a header cookie.
    - This allows for many clusters to be present, while the web UI will only be filtered down to one 
- Dynamic discovery: a discovery type 'json' has been added so that discovery can be done via in-memory json

**Example (multi vtctlds and vtgates)**
with_explanation@2x

### VTAdmin Web
Dynamic discovery relies on a base64-encoded json cookie `cluster` that looks almost exactly like the static file json:
**Raw JSON**
```
{"name": "planetscale-main-main", "vtctlds": [{"host":{"fqdn": "localhost:15000", "hostname": "localhost:15999"}}], "vtgates": [{"host": {"hostname": "localhost:15991"}}]}
```

**Encoded form**
```
eyJuYW1lIjogInBsYW5ldHNjYWxlLW1haW4tbWFpbiIsICJ2dGN0bGRzIjogW3siaG9zdCI6eyJmcWRuIjogImxvY2FsaG9zdDoxNTAwMCIsICJob3N0bmFtZSI6ICJsb2NhbGhvc3Q6MTU5OTkifX1dLCAidnRnYXRlcyI6IFt7Imhvc3QiOiB7Imhvc3RuYW1lIjogImxvY2FsaG9zdDoxNTk5MSJ9fV19
```
**Necessary cookie format**
Screen Shot 2022-01-24 at 10 48 05 AM

### VTAdmin API
To enable dynamic discovery, the `http-enable-dynamic-discovery` flag must be set when spinning up vtadmin-api:
```

{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Launch Package",
            "type": "go",
            "request": "launch",
            "mode": "auto",
            "program": "${workspaceFolder}/go/cmd/vtadmin/main.go",
            "args": [
                "--addr", ":14200",
                "--http-origin", "http://localhost:3000",
                "--logtostderr",
                "--alsologtostderr",
                "--cluster", "id=local,name=local,discovery=staticfile,discovery-staticfile-path=/Users/francesthai/planetscale/vitess/examples/local/vtadmin/discovery.json",
                "--http-tablet-url-tmpl", "http:\/\/localhost:15{{ .Tablet.Alias.Uid }}",
                "--http-enable-dynamic-clusters", "true",
                "--rbac-config",  "/Users/francesthai/planetscale/vitess/examples/local/vtadmin/rbac.yaml"
            ],
            "useApiV1": false,
            "dlvLoadConfig": {
                "followPointers": true,
                "maxVariableRecurse": 1,
                "maxStringLen": 64,
                "maxArrayValues": 64,
                "maxStructFields": -1
            }
        }
    ]
}
```

## To do
- [x] Add diagrams, explanations (documentation)
- [x] Add tests for json discovery
- [x] Add tests for ServeHTTP functionality

## Testing
Follow above steps for VTAdmin-API and VTAdmin-Web to enable dynamic clusters for local example.

## Related Issue(s)
N/A - this works best for Planetscale's use case.

## Checklist
- [x] Should this PR be backported? No
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

## Deployment Notes
N/A

### CODE REVIEW

File: go/vt/vtadmin/api.go
Comment (doeg) [inline go/vt/vtadmin/api.go:219]: (This is not a blocking comment, just a high-level (and possibly hare-brained) idea.) :) 

Do you feel as though the dynamic cluster config functionality is wholly specific to PlanetScale's use case? 

Thinking longer term, one thing we could conceivably do is make vtadmin-api's service discovery and/or HTTP request handlers "plugin-able" so that Vitess operators can write their own plug-ins to handle various points in a request's lifecycle. (Put another way, vtadmin-api could expose "hooks" like, I dunno, `onRequest`, which would call out to any plugin(s) that define a handler.) 

We already take a plugin approach with the [authenticator plugin](https://github.com/vitessio/vitess/blob/f2663ec9e54c60445d77545a5387b0130ffd5762/go/vt/vtadmin/rbac/authentication.go#L137). At Slack, our authenticator plugin actually uses cookies too. :) 

The benefit to Vitess operators is that they can write and maintain their own plugins internally, without having to generalize it to where it can be added to VTAdmin. And, for the open-source project, it'd be less code for the community to maintain.

This is, of course, way out of scope for this change + would require a bunch of design work, but could be a cool approach long term. Would love to hear what you think, and if such a scheme would work for your use case.

**Review by doeg — COMMENTED** [review]
Thank you for such a beautifully thorough PR description (and tests)! 😻 Your PRs are always a delight to review. This is a cool solution to the use case you described. 

I have no blocking suggestions, but (before merging) I am curious if there's an opportunity to streamline the discovery JSON bits.

File: go/vt/vtadmin/api.go
Comment (doeg) [reply go/vt/vtadmin/api.go:219]: ... Maybe I'm just thinking of custom middleware rather than "hooks", actually! lol.

File: go/vt/vtadmin/cluster/discovery/discovery_json_test.go
Comment (notfelineit) [reply go/vt/vtadmin/cluster/discovery/discovery_json_test.go:2]: guilty of copy pasta!! I'll update this, but also I'm sad 2 years just zipped by like that 🥲

File: go/vt/vtadmin/cluster/discovery/discovery_json.go
Comment (notfelineit) [reply go/vt/vtadmin/cluster/discovery/discovery_json.go:52]: They are entirely the same - and I am totally down to extract them!!

File: go/vt/vtadmin/api.go
Comment (notfelineit) [reply go/vt/vtadmin/api.go:219]: I think that would work, and it would require a lot of changes for sure :D but it would be more elegant! The reason we went with the ServeHTTP() design is so that we didn't have to change the way the `API` interface works everywhere. But it is very Planetscale-specific right now!

Trying to make this plugin-able is something I'd be v interested in in the future 😃

Comment (doeg) [reply go/vt/vtadmin/api.go:219]: Yeah! This solution is great + the way you slotted it in is very clean. :)

File: go/vt/vtadmin/cluster/discovery/discovery_json.go
Comment (doeg) [reply go/vt/vtadmin/cluster/discovery/discovery_json.go:52]: Cool -- did you want to do it on this branch or in a separate PR? I'm happy to merge this as-is if that's better for your timeline.

Comment (notfelineit) [reply go/vt/vtadmin/cluster/discovery/discovery_json.go:52]: If I can do it in a different PR, that'd be great!

**Review by doeg — APPROVED** [review]

File: go/vt/vtadmin/cluster/discovery/discovery_json.go
Comment (doeg) [reply go/vt/vtadmin/cluster/discovery/discovery_json.go:52]: For sure, I can merge this in after all the tests pass!

**Pull request merged**
